File: packages/mcp/src/data-context.ts
import { z } from 'zod';
import type { FieldType, JsonObject, JsonValue } from '@metriccanvas/page';

/**
 * 数据上下文快照契约(Schema 元数据 1.0)。
 *
 * 对外契约的三件事都在本模块完成,不得在消费方再写一份:
 * - 形状声明:下方 interface,与 docs/schema-metadata.schema.json 逐层对应
 *   (等价由 tests/data-context-contract.test.ts 的 ajv 守护测试保证);
 * - 结构校验:parseDataContextSnapshot,消费方不得以双重 cast 绕过;
 * - 语义面投影:semanticSurfaceOf,把快照投影为按业务域组织的检索面,
 *   受控句式(取值域)只在这里解析一次;敏感字段在投影与检索侧统一标注,
 *   其取值域语义不外泄(#80)。
 */

export interface DataContextSnapshot {
  formatVersion: '1.0';
  id: string;
  version: string;
  generatedAt: string;
  source: string;
  executionEnvironments: ExecutionEnvironment[];
}

export interface ExecutionEnvironment {
  id: string;
  name: string;
  language: 'dqe';
  endpointRef: string;
  description?: string;
  schemas: DataSchema[];
  constraints: {
    readOnly: true;
    maxRows: number;
    maxColumns: number;
    maxQueriesPerBatch: number;
    timeoutMs: number;
  };
  security: {
    scope: string;
    notes?: string[];
  };
}

export interface DataSchema {
  id: string;
  name: string;
  description: string;
  objects: DataObject[];
  relationships: DataRelationship[];
  verifiedQueries: VerifiedQuery[];
}

export interface DataObject {
  id: string;
  name: string;
  kind: 'dataset';
  description: string;
  fields: DataField[];
}

/** 字段角色提示;与 JSON Schema 的 roleHints 枚举一致('detail' 不属于快照契约)。 */
export type DataFieldRoleHint = 'dimension' | 'measure' | 'time';

export interface DataField {
  name: string;
  type: FieldType;
  description: string;
  aliases?: string[];
  roleHints: DataFieldRoleHint[];
  unit?: string;
  granularity?: string;
  nullable: boolean;
  sensitive: boolean;
}

export interface DataRelationship {
  id: string;
  from: { object: string; field: string };
  to: { object: string; field: string };
  cardinality: 'one-to-one' | 'one-to-many' | 'many-to-one';
  description: string;
}

export interface VerifiedQuery {
  id: string;
  question: string;
  description: string;
  language: 'dqe';
  body: { dsl_list: [JsonObject] };
  resultFields: Array<{
    name: string;
    type: FieldType;
    /** 与 JSON Schema 的 role 枚举一致:快照结果字段只有维度与度量。 */
    role: 'dimension' | 'measure';
    unit?: string;
    nullable: boolean;
  }>;
}

export interface DataContextProvider {
  current(): Promise<DataContextSnapshot>;
}

/* ---------- 结构校验:快照的唯一解析入口 ---------- */

const nonEmptyStringZ = z.string().min(1);
const fieldTypeZ = z.enum(['string', 'number', 'boolean', 'date', 'datetime']);

function uniqueItems(values: readonly string[]): boolean {
  return new Set(values).size === values.length;
}

const dataFieldZ = z
  .strictObject({
    name: nonEmptyStringZ,
    type: fieldTypeZ,
    description: nonEmptyStringZ,
    aliases: z
      .array(nonEmptyStringZ)
      .refine(uniqueItems, { message: '别名不得重复' })
      .optional(),
    roleHints: z
      .array(z.enum(['dimension', 'measure', 'time']))
      .min(1)
      .refine(uniqueItems, { message: '角色提示不得重复' }),
    unit: nonEmptyStringZ.optional(),
    granularity: nonEmptyStringZ.optional(),
    nullable: z.boolean(),
    sensitive: z.boolean()
  });

const relationshipEndZ = z.strictObject({
  object: nonEmptyStringZ,
  field: nonEmptyStringZ
});

const relationshipZ = z.strictObject({
  id: nonEmptyStringZ,
  from: relationshipEndZ,
  to: relationshipEndZ,
  cardinality: z.enum(['one-to-one', 'one-to-many', 'many-to-one']),
  description: nonEmptyStringZ
});

const resultFieldZ = z.strictObject({
  name: nonEmptyStringZ,
  type: fieldTypeZ,
  role: z.enum(['dimension', 'measure']),
  unit: nonEmptyStringZ.optional(),
  nullable: z.boolean()
});

const jsonValueZ: z.ZodType<JsonValue> = z.lazy(() =>
  z.union([
    z.string(),
    z.number(),
    z.boolean(),
    z.null(),
    z.array(jsonValueZ),
    z.record(z.string(), jsonValueZ)
  ])
);
const jsonObjectZ: z.ZodType<JsonObject> = z.record(z.string(), jsonValueZ);

const verifiedQueryZ = z.strictObject({
  id: nonEmptyStringZ,
  question: nonEmptyStringZ,
  description: nonEmptyStringZ,
  language: z.literal('dqe'),
  body: z.strictObject({
    dsl_list: z.tuple([jsonObjectZ])
  }),
  resultFields: z.array(resultFieldZ)
});

const dataObjectZ = z.strictObject({
  id: nonEmptyStringZ,
  name: nonEmptyStringZ,
  kind: z.literal('dataset'),
  description: nonEmptyStringZ,
  fields: z.array(dataFieldZ)
});

const dataSchemaZ = z.strictObject({
  id: nonEmptyStringZ,
  name: nonEmptyStringZ,
  description: nonEmptyStringZ,
  objects: z.array(dataObjectZ),
  relationships: z.array(relationshipZ),
  verifiedQueries: z.array(verifiedQueryZ)
});

const executionEnvironmentZ = z.strictObject({
  id: nonEmptyStringZ,
  name: nonEmptyStringZ,
  language: z.literal('dqe'),
  endpointRef: nonEmptyStringZ,
  description: nonEmptyStringZ.optional(),
  schemas: z.array(dataSchemaZ),
  constraints: z.strictObject({
    readOnly: z.literal(true),
    maxRows: z.int().min(1),
    maxColumns: z.int().min(1),
    maxQueriesPerBatch: z.int().min(1),
    timeoutMs: z.int().min(1)
  }),
  security: z.strictObject({
    scope: nonEmptyStringZ,
    notes: z.array(nonEmptyStringZ).optional()
  })
});

const dataContextSnapshotZ = z.strictObject({
  formatVersion: z.literal('1.0'),
  id: nonEmptyStringZ,
  version: nonEmptyStringZ,
  generatedAt: z.iso.datetime({ offset: true }),
  source: nonEmptyStringZ,
  executionEnvironments: z.array(executionEnvironmentZ)
});

export interface DataContextSnapshotParseError {
  /** JSON Pointer 风格路径,如 /executionEnvironments/0/schemas/1/objects/0/fields/2/name。 */
  path: string;
  message: string;
}

export type DataContextSnapshotParseResult =
  | { ok: true; snapshot: DataContextSnapshot }
  | { ok: false; errors: DataContextSnapshotParseError[] };

/**
 * 数据上下文快照的唯一结构校验入口(#80):形状与
 * docs/schema-metadata.schema.json 等价(守护测试保证),失败给出路径级
 * 错误。消费方一律经由这里进入类型世界,不得 `as unknown as` 硬闯。
 */
export function parseDataContextSnapshot(
  value: unknown
): DataContextSnapshotParseResult {
  const result = dataContextSnapshotZ.safeParse(value);
  if (result.success) {
    // 编译期双向守护:zod 推导形状与导出 interface 互相可赋值,
    // 任何一侧单独改动都会在这里报编译错误。
    const snapshot: DataContextSnapshot = result.data;
    return { ok: true, snapshot };
  }
  return {
    ok: false,
    errors: result.error.issues.map((issue) => ({
      path: `/${issue.path.map(String).join('/')}`,
      message: issue.message
    }))
  };
}

// 编译期反向守护:interface 声明的值必须能通过 zod 推导形状的赋值检查。
type ZodSnapshotShape = z.infer<typeof dataContextSnapshotZ>;
const _interfaceAssignableToZodShape = (
  snapshot: DataContextSnapshot
): ZodSnapshotShape => snapshot;
void _interfaceAssignableToZodShape;

/* ---------- 语义面投影:业务域检索面与受控句式的唯一解析 ---------- */

/**
 * Schema 元数据 1.0 的字段结构封闭,维度取值域以受控句式
 * 「取值域:值1、值2。」写在字段 description(与 DQE 仿真语义面的
 * 同面投影一致)。这是该句式的唯一解析声明,不得另写第二份 pattern。
 */
const VALUE_DOMAIN_PATTERN = /取值域[:：]([^。]+)/u;

export function parseValueDomain(description: string): string[] | undefined {
  const match = VALUE_DOMAIN_PATTERN.exec(description);
  if (!match) return undefined;
  const values = match[1]!
    .split('、')
    .map((value) => value.trim())
    .filter((value) => value !== '');
  return values.length > 0 ? values : undefined;
}

/** 敏感字段标注(#80):字段可被发现,但取值域语义不外泄。 */
function redactSensitiveField(field: DataField): DataField {
  if (!field.sensitive) return field;
  return {
    ...field,
    description: field.description.replace(VALUE_DOMAIN_PATTERN, '取值域:(敏感,已隐去)')
  };
}

/** 业务域摘要:域路由阶段注入模型的全部内容(上下文裁剪,ADR-0037)。 */
export interface BusinessDomainSummary {
  /** 业务域名,即快照中 schema 的 name。 */
  name: string;
  description: string;
}

export interface SemanticSurfaceMetric {
  name: string;
  aliases: string[];
  /** 口径说明(含可加性与时间聚合方式的受控句式原文)。 */
  description: string;
  type: FieldType;
  unit?: string;
  nullable: boolean;
  sensitive: boolean;
}

export interface SemanticSurfaceDimension {
  name: string;
  aliases: string[];
  description: string;
  type: FieldType;
  /** 受控句式解析出的封闭取值域;未声明或敏感字段不投影取值。 */
  values?: string[];
  nullable: boolean;
  sensitive: boolean;
}

export interface SemanticSurfaceTimeDimension {
  name: string;
  aliases: string[];
  description: string;
  type: FieldType;
  /** 该时间维度支持的粒度(granularity 逗号分隔声明)。 */
  granularities: string[];
}

/** 一个业务域的语义面投影:路由命中后注入模型的检索面形状。 */
export interface DomainSemanticSurface {
  businessDomain: string;
  description: string;
  metrics: SemanticSurfaceMetric[];
  dimensions: SemanticSurfaceDimension[];
  timeDimensions: SemanticSurfaceTimeDimension[];
}

/**
 * 把快照投影为按业务域(schema)组织的语义面。V0 两域可全量注入,
 * 形状为多域预留:消费方按命中域取子集,不必一次持有全部投影。
 * 敏感字段保留条目但标注并隐去取值域语义(#80 裁决:标注而非过滤)。
 */
export function semanticSurfaceOf(
  snapshot: DataContextSnapshot
): DomainSemanticSurface[] {
  return snapshot.executionEnvironments.flatMap((environment) =>
    environment.schemas.map((schema) => {
      const surface: DomainSemanticSurface = {
        businessDomain: schema.name,
        description: schema.description,
        metrics: [],
        dimensions: [],
        timeDimensions: []
      };
      for (const object of schema.objects) {
        for (const rawField of object.fields) {
          const field = redactSensitiveField(rawField);
          if (field.roleHints.includes('measure')) {
            surface.metrics.push({
              name: field.name,
              aliases: field.aliases ?? [],
              description: field.description,
              type: field.type,
              ...(field.unit === undefined ? {} : { unit: field.unit }),
              nullable: field.nullable,
              sensitive: field.sensitive
            });
          }
          if (field.roleHints.includes('time')) {
            surface.timeDimensions.push({
              name: field.name,
              aliases: field.aliases ?? [],
              description: field.description,
              type: field.type,
              granularities: (field.granularity ?? '')
                .split(',')
                .map((granularity) => granularity.trim())
                .filter((granularity) => granularity !== '')
            });
          } else if (field.roleHints.includes('dimension')) {
            const values = field.sensitive
              ? undefined
              : parseValueDomain(field.description);
            surface.dimensions.push({
              name: field.name,
              aliases: field.aliases ?? [],
              description: field.description,
              type: field.type,
              ...(values === undefined ? {} : { values }),
              nullable: field.nullable,
              sensitive: field.sensitive
            });
          }
        }
      }
      return surface;
    })
  );
}

export type DataContextMatch =
  | {
      kind: 'environment';
      environmentId: string;
      name: string;
      description?: string;
    }
  | {
      kind: 'schema';
      environmentId: string;
      schemaId: string;
      name: string;
      description: string;
    }
  | {
      kind: 'object';
      environmentId: string;
      schemaId: string;
      objectId: string;
      name: string;
      description: string;
    }
  | {
      kind: 'field';
      environmentId: string;
      schemaId: string;
      objectId: string;
      field: DataField;
    }
  | {
      kind: 'verifiedQuery';
      environmentId: string;
      schemaId: string;
      query: VerifiedQuery;
    };

export interface DataContextSearch {
  current(): Promise<DataContextSnapshot>;
  search(input: { query: string; limit?: number }): Promise<{
    dataContextVersion: string;
    matches: DataContextMatch[];
  }>;
}

export function createDataContextSearch(
  provider: DataContextProvider
): DataContextSearch {
  return {
    current: () => provider.current(),
    async search({ query, limit = 10 }) {
      const snapshot = await provider.current();
      const needle = query.trim().toLocaleLowerCase();
      if (!needle || limit <= 0) {
        return { dataContextVersion: snapshot.version, matches: [] };
      }
      const candidates: Array<{ match: DataContextMatch; text: string }> = [];
      for (const environment of snapshot.executionEnvironments) {
        candidates.push({
          match: {
            kind: 'environment',
            environmentId: environment.id,
            name: environment.name,
            ...(environment.description ? { description: environment.description } : {})
          },
          text: [environment.id, environment.name, environment.description].join(' ')
        });
        for (const schema of environment.schemas) {
          candidates.push({
            match: {
              kind: 'schema',
              environmentId: environment.id,
              schemaId: schema.id,
              name: schema.name,
              description: schema.description
            },
            text: [schema.id, schema.name, schema.description].join(' ')
          });
          for (const object of schema.objects) {
            candidates.push({
              match: {
                kind: 'object',
                environmentId: environment.id,
                schemaId: schema.id,
                objectId: object.id,
                name: object.name,
                description: object.description
              },
              text: [object.id, object.name, object.description].join(' ')
            });
            for (const rawField of object.fields) {
              // 敏感字段标注(#80):检索可发现字段,但匹配文本与返回内容
              // 都使用隐去取值域语义的投影,敏感取值不进入检索面。
              const field = redactSensitiveField(rawField);
              candidates.push({
                match: {
                  kind: 'field',
                  environmentId: environment.id,
                  schemaId: schema.id,
                  objectId: object.id,
                  field
                },
                text: [field.name, field.description, ...(field.aliases ?? [])].join(' ')
              });
            }
          }
          for (const verifiedQuery of schema.verifiedQueries) {
            candidates.push({
              match: {
                kind: 'verifiedQuery',
                environmentId: environment.id,
                schemaId: schema.id,
                query: verifiedQuery
              },
              text: [
                verifiedQuery.id,
                verifiedQuery.question,
                verifiedQuery.description
              ].join(' ')
            });
          }
        }
      }
      return {
        dataContextVersion: snapshot.version,
        matches: candidates
          .map((candidate) => ({
            ...candidate,
            score: matchScore(needle, candidate.text)
          }))
          .filter(({ score }) => Number.isFinite(score))
          .sort((left, right) => left.score - right.score)
          .slice(0, limit)
          .map(({ match }) => match)
      };
    }
  };
}

function matchScore(needle: string, text: string): number {
  const normalized = text.toLocaleLowerCase();
  if (normalized === needle) return 0;
  if (normalized.startsWith(needle)) return 1;
  if (normalized.includes(needle)) return 2;
  return Number.POSITIVE_INFINITY;
}

File: metriccanvas-authoring/tool/metriccanvas_authoring/data/data_context.py
from __future__ import annotations

import json
import re
from copy import deepcopy
from dataclasses import dataclass
from typing import Any, Mapping, Sequence

from jsonschema import Draft202012Validator, FormatChecker
from jsonschema.exceptions import ValidationError

from metriccanvas_authoring.runtime_assets import bundle_root


BUNDLE_ROOT = bundle_root()
DATA_CONTEXT_SCHEMA = BUNDLE_ROOT / "contract-snapshot" / "data-context" / "schema.json"
VALUE_DOMAIN_PATTERN = re.compile(r"取值域[:：]([^。]+)")


@dataclass(frozen=True, slots=True)
class DataContextIssue:
    code: str
    path: str
    message: str


@dataclass(frozen=True, slots=True)
class SemanticMetric:
    name: str
    field_type: str
    unit: str | None
    nullable: bool
    sensitive: bool


@dataclass(frozen=True, slots=True)
class SemanticDimension:
    name: str
    field_type: str
    nullable: bool
    sensitive: bool
    values: tuple[str, ...] | None
    granularities: tuple[str, ...]
    is_time: bool


@dataclass(frozen=True, slots=True)
class MetricEntry:
    business_domain: str
    name: str
    definition: str
    aliases: tuple[str, ...]
    unit: str | None


@dataclass(frozen=True, slots=True)
class DimensionEntry:
    business_domain: str
    name: str
    definition: str
    aliases: tuple[str, ...]
    values: tuple[str, ...] | None
    granularities: tuple[str, ...]
    is_time: bool


@dataclass(frozen=True, slots=True)
class SemanticSurface:
    business_domain: str
    metrics_by_name: Mapping[str, SemanticMetric]
    dimensions_by_name: Mapping[str, SemanticDimension]

    def metric(self, name: str) -> SemanticMetric | None:
        return self.metrics_by_name.get(name)

    def dimension(self, name: str) -> SemanticDimension | None:
        return self.dimensions_by_name.get(name)


@dataclass(frozen=True, slots=True)
class SearchCandidate:
    match: Mapping[str, Any]
    text: str


@dataclass(frozen=True, slots=True)
class DataContext:
    version: str
    surfaces_by_domain: Mapping[str, SemanticSurface]
    metric_entries: tuple[MetricEntry, ...]
    dimension_entries: tuple[DimensionEntry, ...]
    search_candidates: tuple[SearchCandidate, ...]

    def surface(self, business_domain: str) -> SemanticSurface | None:
        return self.surfaces_by_domain.get(business_domain)

    def search(self, query: str, limit: int = 10) -> tuple[Mapping[str, Any], ...]:
        needle = query.strip().casefold()
        if not needle or limit <= 0:
            return ()
        ranked = [
            (score, index, candidate)
            for index, candidate in enumerate(self.search_candidates)
            if (score := _match_score(needle, candidate.text)) is not None
        ]
        ranked.sort(key=lambda item: (item[0], item[1]))
        return tuple(
            deepcopy(candidate.match) for _, _, candidate in ranked[:limit]
        )


def parse_data_context(
    value: Any,
) -> tuple[DataContext | None, tuple[DataContextIssue, ...]]:
    """Validate the neutral snapshot and project its authoring semantic surface."""
    schema = json.loads(DATA_CONTEXT_SCHEMA.read_text(encoding="utf-8"))
    validator = Draft202012Validator(schema, format_checker=FormatChecker())
    issues = [
        DataContextIssue("DATA_CONTEXT_SCHEMA_ERROR", path, error.message)
        for error in validator.iter_errors(value)
        for path in _error_paths(error)
    ]
    if issues:
        return None, tuple(sorted(issues, key=lambda issue: (issue.path, issue.message)))

    snapshot = _mapping(value)
    surfaces: dict[str, SemanticSurface] = {}
    metric_entries: list[MetricEntry] = []
    dimension_entries: list[DimensionEntry] = []
    search_candidates: list[SearchCandidate] = []
    for raw_environment in _sequence(snapshot["executionEnvironments"]):
        environment = _mapping(raw_environment)
        environment_id = str(environment["id"])
        search_candidates.append(
            SearchCandidate(
                match={
                    "kind": "environment",
                    "environmentId": environment_id,
                    "name": environment["name"],
                    **(
                        {}
                        if environment.get("description") is None
                        else {"description": environment["description"]}
                    ),
                },
                text=_join_text(
                    environment["id"],
                    environment["name"],
                    environment.get("description"),
                ),
            )
        )
        for raw_schema in _sequence(environment["schemas"]):
            data_schema = _mapping(raw_schema)
            surface = _project_surface(data_schema)
            surfaces.setdefault(surface.business_domain, surface)
            metric_entries.extend(_metric_entries_for_schema(data_schema))
            dimension_entries.extend(_dimension_entries_for_schema(data_schema))
            search_candidates.extend(
                _search_candidates_for_schema(environment_id, data_schema)
            )
    return (
        DataContext(
            version=str(snapshot["version"]),
            surfaces_by_domain=surfaces,
            metric_entries=tuple(metric_entries),
            dimension_entries=tuple(dimension_entries),
            search_candidates=tuple(search_candidates),
        ),
        (),
    )


def _metric_entries_for_schema(schema: Mapping[str, Any]) -> list[MetricEntry]:
    business_domain = str(schema["name"])
    return [
        MetricEntry(
            business_domain=business_domain,
            name=str(metric["name"]),
            definition=str(metric["description"]),
            aliases=tuple(str(alias) for alias in _sequence(metric.get("aliases", []))),
            unit=_optional_string(metric.get("unit")),
        )
        for raw_metric in _sequence(schema["metrics"])
        for metric in [_mapping(raw_metric)]
    ]


def _dimension_entries_for_schema(
    schema: Mapping[str, Any],
) -> list[DimensionEntry]:
    business_domain = str(schema["name"])
    entries: list[DimensionEntry] = []
    for raw_object in _sequence(schema["objects"]):
        data_object = _mapping(raw_object)
        for raw_field in _sequence(data_object["fields"]):
            field = _mapping(raw_field)
            role_hints = {str(value) for value in _sequence(field["roleHints"])}
            is_time = "time" in role_hints
            if not is_time and "dimension" not in role_hints:
                continue
            entries.append(
                DimensionEntry(
                    business_domain=business_domain,
                    name=str(field["name"]),
                    definition=str(field["description"]),
                    aliases=tuple(
                        str(alias) for alias in _sequence(field.get("aliases", []))
                    ),
                    values=(
                        None
                        if bool(field["sensitive"]) or is_time
                        else _parse_value_domain(str(field["description"]))
                    ),
                    granularities=(
                        tuple(
                            part.strip()
                            for part in str(field.get("granularity", "")).split(",")
                            if part.strip()
                        )
                        if is_time
                        else ()
                    ),
                    is_time=is_time,
                )
            )
    return entries


def _project_surface(schema: Mapping[str, Any]) -> SemanticSurface:
    metrics: dict[str, SemanticMetric] = {}
    for raw_metric in _sequence(schema["metrics"]):
        declaration = _mapping(raw_metric)
        metric = SemanticMetric(
            name=str(declaration["name"]),
            field_type=str(declaration["type"]),
            unit=_optional_string(declaration.get("unit")),
            nullable=bool(declaration["nullable"]),
            sensitive=bool(declaration["sensitive"]),
        )
        _index_name_and_aliases(metrics, metric, declaration)

    dimensions: dict[str, SemanticDimension] = {}
    for raw_object in _sequence(schema["objects"]):
        data_object = _mapping(raw_object)
        for raw_field in _sequence(data_object["fields"]):
            declaration = _mapping(raw_field)
            role_hints = {str(value) for value in _sequence(declaration["roleHints"])}
            is_time = "time" in role_hints
            if not is_time and "dimension" not in role_hints:
                continue
            dimension = SemanticDimension(
                name=str(declaration["name"]),
                field_type=str(declaration["type"]),
                nullable=bool(declaration["nullable"]),
                sensitive=bool(declaration["sensitive"]),
                values=(
                    None
                    if bool(declaration["sensitive"])
                    else _parse_value_domain(str(declaration["description"]))
                ),
                granularities=(
                    tuple(
                        part.strip()
                        for part in str(declaration.get("granularity", "")).split(",")
                        if part.strip()
                    )
                    if is_time
                    else ()
                ),
                is_time=is_time,
            )
            _index_name_and_aliases(dimensions, dimension, declaration)

    return SemanticSurface(
        business_domain=str(schema["name"]),
        metrics_by_name=metrics,
        dimensions_by_name=dimensions,
    )


def _index_name_and_aliases(
    index: dict[str, Any],
    item: SemanticMetric | SemanticDimension,
    declaration: Mapping[str, Any],
) -> None:
    index.setdefault(item.name, item)
    for alias in _sequence(declaration.get("aliases", [])):
        index.setdefault(str(alias), item)


def _parse_value_domain(description: str) -> tuple[str, ...] | None:
    match = VALUE_DOMAIN_PATTERN.search(description)
    if match is None:
        return None
    values = tuple(value.strip() for value in match.group(1).split("、") if value.strip())
    return values or None


def _search_candidates_for_schema(
    environment_id: str, schema: Mapping[str, Any]
) -> list[SearchCandidate]:
    schema_id = str(schema["id"])
    candidates = [
        SearchCandidate(
            match={
                "kind": "schema",
                "environmentId": environment_id,
                "schemaId": schema_id,
                "name": schema["name"],
                "description": schema["description"],
            },
            text=_join_text(schema["id"], schema["name"], schema["description"]),
        )
    ]
    for raw_metric in _sequence(schema["metrics"]):
        metric = _mapping(raw_metric)
        candidates.append(
            SearchCandidate(
                match={
                    "kind": "metric",
                    "environmentId": environment_id,
                    "schemaId": schema_id,
                    "metric": deepcopy(dict(metric)),
                },
                text=_join_text(
                    metric["name"],
                    metric["description"],
                    *_sequence(metric.get("aliases", [])),
                ),
            )
        )
    for raw_object in _sequence(schema["objects"]):
        data_object = _mapping(raw_object)
        object_id = str(data_object["id"])
        candidates.append(
            SearchCandidate(
                match={
                    "kind": "object",
                    "environmentId": environment_id,
                    "schemaId": schema_id,
                    "objectId": object_id,
                    "name": data_object["name"],
                    "description": data_object["description"],
                },
                text=_join_text(
                    data_object["id"],
                    data_object["name"],
                    data_object["description"],
                ),
            )
        )
        for raw_field in _sequence(data_object["fields"]):
            field = _redacted_field(_mapping(raw_field))
            candidates.append(
                SearchCandidate(
                    match={
                        "kind": "field",
                        "environmentId": environment_id,
                        "schemaId": schema_id,
                        "objectId": object_id,
                        "field": field,
                    },
                    text=_join_text(
                        field["name"],
                        field["description"],
                        *_sequence(field.get("aliases", [])),
                    ),
                )
            )
    for raw_query in _sequence(schema["verifiedQueries"]):
        query = _mapping(raw_query)
        candidates.append(
            SearchCandidate(
                match={
                    "kind": "verifiedQuery",
                    "environmentId": environment_id,
                    "schemaId": schema_id,
                    "query": deepcopy(dict(query)),
                },
                text=_join_text(
                    query["id"], query["question"], query["description"]
                ),
            )
        )
    return candidates


def _redacted_field(field: Mapping[str, Any]) -> dict[str, Any]:
    result = deepcopy(dict(field))
    if bool(result["sensitive"]):
        result["description"] = VALUE_DOMAIN_PATTERN.sub(
            "取值域:(敏感,已隐去)", str(result["description"])
        )
    return result


def _match_score(needle: str, text: str) -> int | None:
    normalized = text.casefold()
    if normalized == needle:
        return 0
    if normalized.startswith(needle):
        return 1
    if needle in normalized:
        return 2
    return None


def _join_text(*values: object) -> str:
    return " ".join("" if value is None else str(value) for value in values)


def _error_paths(error: ValidationError) -> list[str]:
    base = _pointer(error.absolute_path)
    if error.validator == "required" and isinstance(error.instance, dict):
        missing = [name for name in error.validator_value if name not in error.instance]
        return [_join_pointer(base, name) for name in missing]
    return [base]


def _join_pointer(base: str, part: object) -> str:
    encoded = str(part).replace("~", "~0").replace("/", "~1")
    return f"{base}/{encoded}" if base else f"/{encoded}"


def _pointer(parts: Any) -> str:
    result = ""
    for part in parts:
        result = _join_pointer(result, part)
    return result


def _mapping(value: object) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise TypeError("expected object")
    return value


def _sequence(value: object) -> Sequence[object]:
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)):
        raise TypeError("expected array")
    return value


def _optional_string(value: object) -> str | None:
    return value if isinstance(value, str) else None
